File: Easy_Problems/4. Tournament Winner/Javascript/Solution.js
// O(n) time | O(k) space where k is the number of teams
// Hash Map Approach + One Pass Approach

const HOME_TEAM_WON = 1;
const POINTS = 3;

function findTournamentWinner(competition_arr, results_arr) {
    let currBestTeam = "";
    const scores = { [currBestTeam]: 0 };

    for (let i = 0; i < competition_arr.length; i++) {
        const [homeTeam, awayTeam] = competition_arr[i];
        const result = result_arr[i];

        const winningTeam = result === HOME_TEAM_WON ? homeTeam : awayTeam;

        updateScores(scores, winningTeam);

        if (scores[winningTeam] > scores[currBestTeam]) {
            currBestTeam = winningTeam;
        }
    }

    return currBestTeam;
}

function updateScores(scores, team) {
    const prevScore = scores[team] || 0;
    scores[team] = prevScore + POINTS;
}

File: Easy_Problems/4. Tournament Winner/Python/Solution.py
# O(n) time | O(k) space where k is the number of teams
# Hash Map Approach + One Pass Approach

def findTournamentWinner(competition_arr, results_arr):

    HOME_TEAM_WON = 1;
    WIN_POINTS = 3;
    
    current_best_team = "";
    scores = {current_best_team: 0}

    for idx, competition in enumerate(competition_arr):
        [home_team, away_team] = competition
        result = results_arr[idx]

        winningTeam = home_team if result == HOME_TEAM_WON else away_team
        updateScores(scores, winningTeam, WIN_POINTS)

        if scores[winningTeam] > scores[current_best_team]:
            current_best_team = winningTeam
    
    return current_best_team

def updateScores(scores, team, points):
    if team not in scores:
        scores[team] = 0
    scores[team] += points

# Tests for findTournamentWinner
assert(findTournamentWinner([["A", "B"]], [1]) == "A")
assert(findTournamentWinner([["A", "B"]], [0]) == "B")
assert(findTournamentWinner([["HTML", "C#"], ["C#", "Python"], ["Python", "HTML"]], [0,0,1]) == "Python")
